aidial_adapter_openai/utils/reflection.py

```diff
@@ -0,0 +1,44 @@
+import inspect
+from typing import Any, Callable, Coroutine, TypeVar
+
+from aidial_sdk.exceptions import HTTPException as DialException
+
+T = TypeVar("T")
+
+
+async def call_with_extra_body(
+    func: Callable[..., Coroutine[Any, Any, T]], arg: dict
+) -> T:
+    if has_kwargs_argument(func):
+        return await func(**arg)
+
+    expected_args = set(inspect.signature(func).parameters.keys())
+    actual_args = set(arg.keys())
+
+    extra_args = actual_args - expected_args
+
+    if extra_args and "extra_body" not in expected_args:
+        raise DialException(
+            f"Extra arguments aren't supported: {extra_args}.",
+            400,
+            "invalid_request_error",
+        )
+
+    arg["extra_body"] = arg.get("extra_body") or {}
+
+    for extra_arg in extra_args:
+        arg["extra_body"][extra_arg] = arg[extra_arg]
+        del arg[extra_arg]
+
+    return await func(**arg)
+
+
+def has_kwargs_argument(func: Callable[..., Coroutine[Any, Any, Any]]) -> bool:
+    """
+    Determines if the given function accepts a variable keyword argument (**kwargs).
+    """
+    signature = inspect.signature(func)
```

Comment:
Minor:

Theoretically we could wrap this in inspect.signature in something like

```
@functools.lru_cache(maxsize=64)
def _get_signature(func):
    return inspect.signature(func)
```

Because we basically always call it twice in `call_with_extra_body`


But it would be super-minor optimization, since it's fast function call

Reply:
yeah, it's a micro-optimization, let's not go there just yet

Reply:
Agreed